float8_experimental/fsdp_utils.py

```diff
@@ -64,7 +64,9 @@ def precompute_float8_dynamic_scale_for_fsdp(module: nn.Module) -> None:
         scale_tensor = torch.clamp(scale_tensor, max=torch.finfo(torch.float16).max)
     scales = torch.split(scale_tensor, 1)  # Replicate
     for scale, float8_linear in zip(scales, float8_linears):
-        float8_linear.weight._local_tensor._precomputed_scale = scale._local_tensor
+        float8_linear.weight._local_tensor._precomputed_scale = (
+            scale._local_tensor.squeeze()
```

Comment:
make sure tensor is like `tensor(4674.8633)` instead of `tensor([[4674.8633]]`

Reply:
Can we check the traces? I want to make sure there is no CPU sync point introduced from making this tensor a scalar tensor.

Reply:
do we know the reasoning for why the current behavior is not supported with compile?  This might not scale long term as we add other scaling granularities like rowwise or blockwise.

Reply:
> Can we check the traces? I want to make sure there is no CPU sync point introduced from making this tensor a scalar tensor

sure. I checked the trace and it's seems to be purely executed on cpu, no kernels launch, and no `cudaStreamSynchronize` if that's what you refer as "CPU sync point"
<img width="289" alt="Screenshot 2024-07-17 at 2 26 24 PM" src="https://github.com/user-attachments/assets/225b6a1c-f567-4cd9-9cf3-47536d275ee0">

Reply:
I mean when the scalar is used later downstream.

Reply:
> do we know the reasoning for why the current behavior is not supported with compile? This might not scale long term as we add other scaling granularities like rowwise or blockwise.

TL;DR: this is more like a bug when I implement `precompute_float8_dynamic_scale_for_fsdp`

for the 1st iteration, `self._precomputed_scale` is `None` and thus we calcuclate scale through `cast_to_float8_e4m3_dynamic`, where scale are in `tensor(4674.8633)`

for the 2nd iteration after `precompute_float8_dynamic_scale_for_fsdp`, we have `self._precomputed_scale=tensor([[4674.8633]])` because I only called `torch.split(scale_tensor, 1)` without `.squeeze`. It should be my responsibility to keep `precompute_float8_dynamic_scale_for_fsdp` consistent with `cast_to_float8_e4m3_dynamic`

Reply:
> I mean when the scalar is used later downstream.

ah, I see. I should be looking for `cudaStreamSynchronize`, right ?

Reply:
I think it would be like `cudaDeviceSynchronize` if I understand correctly (but basically you would see the CPU thread blocked).

Reply:
> I mean when the scalar is used later downstream.

`_precomputed_scale` will be used inside `fsdp_pre_all_gather` when calling

```
float8_tensor = Float8Tensor.to_float8(
    self._tensor,
    self._precomputed_scale,
    torch.float8_e4m3fn,
    mm_config=self._mm_config,
)
```

I annotated the function with `record_function("Float8Tensor.to_float8")`. Here are the snapshots for cpu thread and cuda stream

<img width="318" alt="Screenshot 2024-07-17 at 4 33 53 PM" src="https://github.com/user-attachments/assets/008beb59-68a2-4f6d-bb54-d84ae2cc98b9">

<img width="1302" alt="Screenshot 2024-07-17 at 4 34 31 PM" src="https://github.com/user-attachments/assets/cdcc3c38-4ed1-46b5-92d3-6fa728e0d996">

in both cases, I do not see `cudaStreamSynchronize` and cuda stream stays ahead of cpu thread

any worries ?

Reply:
sounds good! should be fine